src/tests/governance/test_timelock.cairo

```diff
@@ -0,0 +1,1777 @@
+}
+
+#[test]
+#[should_panic(expected: ('Timelock: expected Pending op', 'ENTRYPOINT_FAILED'))]
+fn test_cancel_when_unset() {
+    let (mut timelock, _) = setup_dispatchers();
+    let invalid_id = 0;
+
+    // PROPOSER is also CANCELLER
+    testing::set_contract_address(PROPOSER());
+    timelock.cancel(invalid_id);
+}
+
+#[test]
+#[should_panic(expected: ('Caller is missing role', 'ENTRYPOINT_FAILED'))]
+fn test_cancel_unauthorized() {
+    let (mut timelock, mut target) = setup_dispatchers();
+    let predecessor = NO_PREDECESSOR;
+    let salt = 0;
+    let delay = MIN_DELAY;
+
+    let call = single_operation(target.contract_address);
+    let target_id = timelock.hash_operation(call, predecessor, salt);
+
+    // Schedule
+    testing::set_contract_address(PROPOSER());
+    timelock.schedule(call, predecessor, salt, delay);
+
+    // Cancel
+    testing::set_contract_address(OTHER());
+    timelock.cancel(target_id);
+}
+
+//
+// update_delay
+//
+
+#[test]
+#[should_panic(expected: ('Timelock: unauthorized caller', 'ENTRYPOINT_FAILED'))]
+fn test_update_delay_unauthorized() {
+    let mut timelock = deploy_timelock();
+
+    timelock.update_delay(NEW_DELAY);
+}
+
+#[test]
+fn test_update_delay_scheduled() {
+    let mut timelock = deploy_timelock();
+    let predecessor = NO_PREDECESSOR;
+    let salt = 0;
+    let delay = MIN_DELAY;
+    let event_index = 0;
+
+    let call = Call {
+        to: timelock.contract_address,
+        selector: selector!("update_delay"),
+        calldata: array![NEW_DELAY.into()].span()
+    };
+    let target_id = timelock.hash_operation(call, predecessor, salt);
+
+    // Schedule
+    testing::set_contract_address(PROPOSER());
+    timelock.schedule(call, predecessor, salt, delay);
+    assert_operation_state(timelock, OperationState::Waiting, target_id);
+    assert_only_event_schedule(
+        timelock.contract_address, target_id, event_index, call, predecessor, delay
+    );
+
+    // Fast-forward
+    testing::set_block_timestamp(delay);
+
+    // Execute
+    testing::set_contract_address(EXECUTOR());
+    timelock.execute(call, predecessor, salt);
+    assert_operation_state(timelock, OperationState::Done, target_id);
+    assert_event_delay(timelock.contract_address, MIN_DELAY, NEW_DELAY);
+    assert_only_event_execute(timelock.contract_address, target_id, event_index, call);
+
+    // Check new minimum delay
+    let get_new_delay = timelock.get_min_delay();
+    assert_eq!(get_new_delay, NEW_DELAY);
+}
+
+//
+// Internal
+//
+
+//
+// initializer
+//
+
+#[test]
+fn test_initializer_single_role_and_no_admin() {
+    let mut state = COMPONENT_STATE();
+    let contract_state = CONTRACT_STATE();
+    let min_delay = MIN_DELAY;
+
+    let proposers = array![PROPOSER()].span();
+    let executors = array![EXECUTOR()].span();
+    let admin_zero = ZERO();
+
+    state.initializer(min_delay, proposers, executors, admin_zero);
+    assert!(contract_state.has_role(DEFAULT_ADMIN_ROLE, admin_zero));
```

Comment:
Should we set a contract_address different from zero here since the initializer should not assign a role when the admin param is zero? For consistency.

Reply:
Yeah, we can do that. I think it's worth adding another test specifically checking that the initializer doesn't set the admin when it's zero. Will update